### ISSUE TITLE
Store sensitive data as credentials

### ISSUE THREAD

**Issue by Caprico85** [mikejac/node-red-contrib-google-smarthome#26]
Node-RED allows to store sensitive data as credentials. See https://nodered.org/docs/creating-nodes/credentials. Credentials are encrypted by Node-RED. And much more important, credentials are not exported if you export a flow.

Currently, usernames, passwords, secrets etc. are stored as normal config fields. If you export a flow containing a Google Home node, your sensitive data is exported with it. If you don't replace that data in the exported file, an attacker already has lots of the data needed to control your home.

I don't know if we can automatically migrate existing data to credentials. Otherwise the user would have to reenter all of his/her data after the upgrade. If anyone knows more, please answer.

**Comment by mikejac** [comment]
I'm pretty sure we can't automatically migrate data. The facility to store credentials is enabled in the call to 'RED.nodes.registerType()' as far as I recall.

**Comment by mikejac** [comment]
I think this should wait until we can go to version 0.1.0 (which I don't feel we can yet - the inline node documentation is not in place for all nodes).

**Comment by mikejac** [comment]
Hmmm, I think it would be possible for the GoogleSmartHomeNode() to use either the existing credentials _or_ Node-RED secure credentials (by testing for the existance of 'this.credentials.username').

**Comment by mikejac** [comment]
Not that easy. Problem is in the HTML file - the 'this.credentials.password' isn't accessable. So I would say not possible to do automatic migration :-(

**Comment by Caprico85** [comment]
> the 'this.credentials.password' isn't accessable

this.credentials.password isn't accessible. But you can add an `oneditprepare` function that sets the form field using `$("#node-config-input-password").val(...)`. The old unsecure value can still be found in `this.password`. When the config is saved and deployed, the field is now saved as a secure credential field. Tested this and it works.

Unfortunately, the old, unsecured value is still present in flows.json and is still being exported. Still need to find a way to remove it from there.

**Comment by mikejac** [comment]
Would this be possible: 1. In the HTML-part, `node-config-input-username` and `node-config-input-password` 'points' to the secure credentials. 2. `oneditprepare` populates those fields if they are empty (getting values from the non-secure fields) and then clears `this.username` and `this.password` ?

**Comment by Caprico85** [comment]
Populating node-config-input-username with the non-secure field is possible.

I haven't yet found a way to clear this.username and this.password. Simply setting them to null or undefined doesn't help, neither in HTML nor JS files. These changes are not written back to flows.json or Node-Reds internal storage. But maybe I called that in the wrong place. Maybe there is a save() function that needs to be called. Maybe we can hook into Node-Reds load/save logic to prevent the old values from being loaded or written to disk. If all else fails, renaming the fields might also be an option. There are still some options I want to try out.

**Comment by Caprico85** [comment]
After trying for hours and hours I realized there are two calls to registerType() (in the JS and the HTML file). I only changed the call to registerType() in the HTML file. After fixing that, old config fields are correctly removed on deploying. Only problem is that I can't access the old config fields in the HTML file anymore to migrate them. Next try was migrating the configs directly in the JS file when the node is registered. Works great, except the data is not saved to disk and is lost when Node-RED is restarted. We [can't save the credentials from code](https://discourse.nodered.org/t/change-node-credentials-at-runtime/19590). and we can't rely on the user to save the credentials in the UI. Even more annoying is that your flows_cred.json is encrypted. No chance to see the saved credentials or removing them to try migration again.

So currently I'm out of ideas and I give up (for now) ...
